### PR TITLE
JAX-RS metrics support, using Servlet API + Update to MP Metrics 2.3 Final

### PR DESCRIPTION
**Pull request by jmartisk**
Fixes https://github.com/smallrye/smallrye-metrics/issues/231

### CODE REVIEW

**Comment by Ladicek** [comment]
Suggestion: what about a counter of all successful and failed requests? Or even a histogram of status codes?

**Comment by jmartisk** [comment]
This is somewhat funny stuff that was quite tough to come up with, I'd appreciate if you guys could have a look if this makes any sense to you (@Ladicek , @kenfinnigan ). See the Javadoc for explanation why I had to use both a servlet filter and JAX-RS request filter.
Quarkus in Vert.x mode (that means without servlets) will use a little bit nicer way to do this: a JAX-RS request filter combined with a Vert.x response hook - see https://github.com/jmartisk/quarkus/commit/0a932a8d9f9769097c155ecff0bea10ae307b7d9#diff-59c7a0f264a275f199b9633ef6fb8c6b

 If you have a better idea how to solve this, I'm all ears :D  This has been baffling me for quite some time.

**Comment by jmartisk** [comment]
> Suggestion: what about a counter of all successful and failed requests? Or even a histogram of status codes?

That's a fine point, we could do that, this is for start to implement what is in the specification. Let's first decide if this approach is feasible and then we can add more stuff. The problem with that will be that our vendor-specific metrics will need to be in the `vendor` scope and this is in `base` scope, but that's just how it is...

**Comment by Ladicek** [comment]
Ah, I had no idea there's a spec for this now :-)

**Comment by jmartisk** [comment]
And regarding tests: TCK tests for this are still not merged in MP Metrics repo, hopefully they will be soon (I'm not doing them). I will probably also add our own smoke test for this into SR Metrics.

File: implementation/src/main/java/io/smallrye/metrics/jaxrs/JaxRsMetricsServletFilter.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+    @Override
+    public void init(FilterConfig filterConfig) {
+    }
+
+    @Override
+    public void doFilter(ServletRequest servletRequest, ServletResponse servletResponse, FilterChain filterChain)
+            throws IOException, ServletException {
+        long start = System.nanoTime();
+        filterChain.doFilter(servletRequest, servletResponse);
```
Comment (Ladicek) [inline implementation/src/main/java/io/smallrye/metrics/jaxrs/JaxRsMetricsServletFilter.java]: `try`/`finally` around this?

**Comment by Ladicek** [comment]
It looks fine to me. The only question that comes to mind is about Quarkus, where sometimes JAX-RS can run on top of servlet and sometimes on pure Vert.x, but it seems you've got that one covered.

File: implementation/src/main/java/io/smallrye/metrics/jaxrs/JaxRsMetricsServletFilter.java (hunk 1 above)
Comment (jmartisk) [reply implementation/src/main/java/io/smallrye/metrics/jaxrs/JaxRsMetricsServletFilter.java]: Yeah, good catch

**Comment by kenfinnigan** [comment]
So this requirement is part of the Metrics spec? And the only way for an implementation to support it is to use a Servlet Filter?

If so, I'd say that's a problem as MP shouldn't be developing specs that require implementations to use non MP specs, ie. servlet

**Comment by Ladicek** [comment]
I'm pretty sure it would be possible to implement the spec simply by modifying the JAX-RS implementation, so technically the spec doesn't require using the servlet API. But if we don't want to modify our JAX-RS implementation, we naturally use whatever our JAX-RS implementation sits on top of, which can be servlet or Vert.x. At least that's my understanding.

**Comment by kenfinnigan** [comment]
My interpretation from the javadoc comment in the PR was that it's a workaround for limitations in the JAX-RS spec.

If it's actually RESTEasy that only has the issue, then yes we can workaround with Servlet API.

If it's an issue with JAX-RS spec, we can't work around with Servlet API and MP Metrics needs to be modified

**Comment by jmartisk** [comment]
I would say it's a problem of the JAX-RS spec, because it doesn't explicitly require something to work, and so in RESTeasy it just doesn't work. In some other implementations it could, I don't know.

I agree this is quite unfortunate, and I initially kinda fought back against it being in the spec (see https://github.com/eclipse/microprofile-metrics/issues/293). It is in there now, but it's marked as optional, and while guys said they wanted to make it required in next releases, I think we might want to keep it optional due to this issue, unless someone invents a reasonable solution that can be used with any JAX-RS implementation and without non-MP dependencies.

On the other hand, OpenTracing suffers from pretty much exactly the same problem, has had it for a long time, and it's at the very core of OpenTracing concepts, because a similar workaround is needed to be able to finish open spans for JAX-RS calls.

**Comment by jmartisk** [comment]
Someone from WildFly might also want to have a look at this - @jmesnil ? Does this look suitable for WildFly?

**Comment by jmesnil** [comment]
@ronsigal, could you please review it? How would we integrate that into WildFly? I suppose we can add the Servlet filter in the jaxrs subsystem?

**Comment by ronsigal** [comment]
@jmesnil, yes, I'll take a look.

**Comment by jmartisk** [comment]
I added support for the new TCK module that verifies this (the module is named `optional`) - I tested it with this not-yet-merged PR https://github.com/eclipse/microprofile-metrics/pull/529 and it's passing right now

**Comment by ronsigal** [comment]
Hi everyone,

Sorry for the late response. So, the problem in JAX-RS is that an exception that isn't caught by an ExceptionMapper isn't required to go through the filter chain, right? I can think of two approaches, other than the one given here:

1. Modify the JAX-RS spec to mandate a default ExceptionMapper.
2. Include, in Smallrye Metrics, a default ExceptionMapper, with the lowest possible priority.

Obviously, approach 1 is a long term solution, although it seems reasonable. 

I can think of only one problem with approach 2. Given that a Smallrye Metrics ExceptionMapper is, with respect to a JAX-RS implementation, an application supplied provider, and "an application-supplied provider MUST always be preferred over a pre-packaged one if a single one is
required" [JAX-RS spec, Section 4.1.3], a Smallrye Metrics ExceptionMapper can't have the absolutely lowest priority, and it's presence could interfere with existing behavior. But Smallrye Metrics could have an option of disabling its default ExceptionMapper. [Mmmm, not sure how, offhand.] If the JAX-RS implementation already supplies a default ExceptionMapper, at least an ExceptionMapper exists, and the filter chain would get executed.

WDYT?

-Ron

**Comment by jmartisk** [comment]
Modifying JAX-RS spec of course would simplify things a lot, but yeah that is a long term thing and we need to release this in about 2 weeks :)

While I'm not very familiar with exception mappers, I considered this option as well, but I wasn't sure how exactly the mapper should map an exception so that it will look like no mapping was performed (I didn't want to duplicate logic from RestEasy and basically make this dependent on the JAX-RS implementation). And how to register it so that it has the lowest priority, but is always there. This would need some deeper diving. 
After some thinking, I just felt like the Servlet/Vertx-based solution seemed easier, and I had a relatively clear idea how to do that, so I went ahead with it. So if someone comes up with a pure JAX-RS solution that is less hacky than this, we could give it a try, but we're kinda running out of time before this needs to be released.

**Comment by ronsigal** [comment]
Hey @jmartisk,

Yeah, I didn't want to suggest that I assumed you hadn't thought about ExceptionMappers. ;-)

re: "I wasn't sure how exactly the mapper should map an exception so that it will look like no mapping was performed"

I would just create a Response that holds the text of the Exception and copies the response code.

re: "how to register it so that it has the lowest priority, but is always there."

There are two factors determining priority: "When choosing an exception mapping provider to map an exception, an implementation MUST use the provider whose generic type is the nearest superclass of the exception. If two or more exception providers are applicable, the one with the highest priority MUST be chosen as described in Section 4.1.3. [JAX-RS spec, Section 4.4]"

Anyway, I'm not trying to slow things down. Adding an ExceptionMapper seems simpler to me, but you've already done the hard work. So I'm not objecting.

-Ron

**Comment by jmartisk** [comment]
Thanks @ronsigal . The main concern now is if this looks like it can be reasonably integrated into WildFly. The TCK module (which I've just added today) uses WildFly as the runtime container, but it runs using a very different deployment strategy than WildFly applications will - it ignores the Metrics subsystem and instead brings in its own SmallRye Metrics JAR  inside a deployment. The servlet filter is added through the deployment's `web.xml` and JAX-RS request filter is added through ServiceLoader. So the test basically uses WildFly just as a servlet + JAX-RS container. 
For regular deployments, WildFly itself will probably need to add these two using subsystem code.

**Comment by ronsigal** [comment]
Ok, I think I understand the question now.

I was thinking that it would be enough to annotate JaxRsMetricsServletFilter with @Provider, but the JAX-RS spec says, "All application-supplied providers implement interfaces in the JAX-RS API and MAY be annotated with @Provider for automatic discovery purposes; the integration of pre-packaged providers into the JAX-RS runtime is implementation dependent." I'm going to have to look at the RESTEasy and/or jaxrs subsystem code ... . In fact, the RESTEasy builtin providers are loaded by the service loader mechanism, but I'm not sure if that's necessary ...

**Comment by jmartisk** [comment]
I wanted to avoid annotating the filter with `@Provider` to make sure it won't be picked up automatically when it's not supposed to, this way the container will have more control over enabling it. Not sure if this really could happen (depends a lot on the particular runtime, and we support multiple runtimes), but this solution looks safer. The same holds for the servlet filter, which could be annotated with `@WebFilter`.

**Comment by ronsigal** [comment]
Well, here's a partial, possible, solution. As of Wildfly 19, RESTEasy can be configured in standalone.xml, et. al. For example,
```
<subsystem xmlns="urn:jboss:domain:jaxrs:2.0">
    <resteasy-providers>
        <class>io.smallrye.metrics.jaxrs.JaxRsMetricsFilter</class>
    </resteasy-providers>
</subsystem>
```
Is that appropriate in this context? Still, we have to *find* io.smallrye.metrics.jaxrs.JaxRsMetricsFilter, which might mean a change to the wildfly-jaxrs subsystem. I'll keep looking.

**Comment by jmartisk** [comment]
I was thinking more along the lines that if a deployment processor detects that a deployment contains metrics *and* JAX-RS endpoints *and* the functionality is enabled (by looking at the configuration or a property), it will register the two filters for that deployment..? If that's possible. I don't remember that much about WildFly inner mechanisms.

**Comment by ronsigal** [comment]
@jmartisk, I'll look into that. I'm going to be offline this morning, but I'll be back this afternoon.

**Comment by jmartisk** [comment]
I am moving it from draft because MP Metrics 2.3-RC2 that contains the TCK tests is now out, so the CI checks contain the TCK from now on

**Comment by ronsigal** [comment]
Hey @jmartisk,

Sorry I disappeared. I developed a cold on Tuesday afternoon, and I still feel pretty bad. I'm hoping to have the energy to get back to his issue tomorrow. If not, I'll see if I can find some else to work on it.

-Ron

**Comment by jmartisk** [comment]
Spec 2.3 Final is now out, so this PR now contains the upgrade as well

**Comment by jmartisk** [comment]
Hey @ronsigal , any updates please? This is (I hope!) the last thing we need to merge before releasing 2.4.0. I just wanted to do a quick check that the WildFly team will not shout at me later that I implemented something that's not possible to integrate.

**Comment by ronsigal** [comment]
Hi @jmartisk,

The fact is, I'm still sick. When it became clear to me on Monday that I just didn't have the clarity to think about this issue, I asked another one of our RESTEasy people, Rebecca Searles, if she would look into it. Rebecca has experience with both RESTEasy and Wildfly, so I think she's a good choice. I should have said something sooner, though. Sorry for disappearing again.

By the way, I was thinking that one solution could involve building a kludgy test into RESTEasy testing for an orphan exception in the context metrics and just running the matching container filter. I haven't tried it, though.

-Ron

**Comment by ronsigal** [comment]
re; "something that's not possible to integrate."

Well, the important thing is to be able to detect a metrics situation and inject the appropriate filters. With respect to RESTEasy, that doesn't seem hard (I say that without having tested it though). The reason I keep leaning towards keeping it all in RESTEasy is just that I don't have experience with the analogous situation for servlets.

**Comment by rsearls** [comment]
I am replacing Ron in this task.   I have a proposed change for Resteasy that will allow the execution of both a ContainerRequestFilter and ContainerResponseFilter.  Does providing is
negate the need for a speciality servlet?

How are you planning to get your filter(s) inserted into jaxrs apps?

**Comment by jmartisk** [comment]
> I am replacing Ron in this task. I have a proposed change for Resteasy that will allow the execution of both a ContainerRequestFilter and ContainerResponseFilter. Does providing is
> negate the need for a speciality servlet?

Basically yes, but it depends when such change will be available, I suppose it will take some time before this Resteasy change can land in WildFly. Especially since it's a change in behavior that some users might find unexpected. I suppose we would keep this old solution (with a servlet filter) present in SmallRye for the time being anyway, and if this Resteasy change indeed happens, it won't be difficult to tune SmallRye a bit to make use of that... It can be done in a micro release of SmallRye Metrics, if you just let me know then I can implement that.

> 
> How are you planning to get your filter(s) inserted into jaxrs apps?

In WildFly you mean? That's what I am basically asking you guys :) I justed wanted to get an ACK from the WildFly team that this can be done in WildFly - that it seems possible to automatically insert these filters into applications, probably by tuning something in the JAXRS *or* Metrics subsystem. I know how to do that for Quarkus and I will do the integration there, but I'm not that familiar with WildFly.

**Comment by jmartisk** [comment]
Anyway, also bear in mind that this is an optional feature per the specification, so if there's a significant problem integrating it and in the end we don't do that, it does not mean WildFly won't be MP compliant.

**Comment by jmesnil** [comment]
@jmartisk In the dev analysis to integrate MP Metrics 2.3 Final, I state that this optional feature will be handled separately from the MP Metrics upgrade.
I don't want to hold the release of smallrye-metrics for it as it can be addressed later.

@rsearls if you think that the integration in WildFly is doable, that's enough for me. We don't have to do the actual integration for MP 2.3.

**Comment by jmartisk** [comment]
Hey @jmesnil , awesome, thanks a lot.
I'll merge this then. Should you guys in WildFly need any assistance with the JAX-RS metrics, I will be available.

**Comment by rsearls** [comment]
> In WildFly you mean? That's what I am basically asking you guys :) I justed wanted to get an ACK from the WildFly team that this can be done in WildFly - that it seems possible to automatically insert these filters into applications, probably by tuning something in the JAXRS or Metrics subsystem. I know how to do that for Quarkus and I will do the integration there, but I'm not that familiar with WildFly.

Is this going to be attached to every jaxrs app deployed through wildfly or is there some flag the user sets in wildfly to indicate it is to be added?

**Comment by jmartisk** [comment]
> Is this going to be attached to every jaxrs app deployed through wildfly or is there some flag the user sets in wildfly to indicate it is to be added?

I don't know, that will be up to WildFly team to decide, but I suppose enabling it by default is not a good idea for performance purposes. At least in Quarkus it will be something that needs to be explicitly enabled by a config property

**Comment by rsearls** [comment]
When do you think the WildFly team will be providing information on this?  Their decision will determine the options I will suggest.

**Comment by jmesnil** [comment]
@rsearls I think we will do like Quarkus does and make it opt-in.
If the user enables it, the microprofile-metrics subsystem would then be in charge of adding the filter to every jaxrs deployment.

**Comment by rsearls** [comment]
@jmesnil can you provide details of how the user will indicate this in wildfly?

**Comment by ronsigal** [comment]
@jmartisk, @jmesnil, @rsearls,

1. Our plan has been to release RESTEasy 3.11.0.Final with this issue and MP REST Client 1.4. so we should be able to get it work for the next WF 19.

2. By the way, do you have a smallrye feature request for this? Can we finesse it to include where happens in RESTEasy?

**Comment by jmesnil** [comment]
> @jmesnil can you provide details of how the user will indicate this in wildfly?

there are different possibilities:
1. an attribute in the `microprofile-metrics-smallrye` subsystem that enables JAX-RS metrics globally for all WAR deployments
2. if we want more granularity, a MP Config Property so that each WAR deployment can specify if WildFly should produce metrics for it.

In both cases, the `microprofile-metrics-smallrye` subsystem should be the one that handles this setting and add the filter to the deployment (most likely by delegating to RESTeasy)

**Comment by rsearls** [comment]
As I understand this then, due to timing smallrye would merge their
current design into wfly-19. RESTEasy 3.11.0.Final will be released with 
MP REST Client 1.4 for wfly-19.  My proposed change will be merged into a 
later version of RESTEasy and smallrye will provide corresponding changes 
in some future version of wfly.

**Comment by ronsigal** [comment]
FYI, I have created https://github.com/eclipse-ee4j/jaxrs-api/issues/855, "Mandate default ExceptionMapper".

**Pull request merged**